### ISSUE TITLE
Git basics

### ISSUE THREAD

**Issue by loucadufault** [loucadufault/neural-network#1]
There's two ways to learn git. One is to understand every inner working of git and all the obscure features, the other is to learn a handful of commands and what to do when stuff messes up and hope for the best. I'm going to teach the second one here, so basically how git works as I understand it, how it interacts with github.

### What is git?
git is a version-control software (VCS) that is usually run from the [unix terminal](https://www.youtube.com/watch?v=zw7Nd67_aFw) (mac and linux) or [command line](https://www.youtube.com/watch?v=uE9WgNr3OjM) (windows). 

"Version control systems are a category of software tools that help a software team manage changes to source code over time. Version control software keeps track of every modification to the code in a special kind of database. If a mistake is made, developers can turn back the clock and compare earlier versions of the code to help fix the mistake while minimizing disruption to all team members."

"Git is an actively maintained open source project originally developed in 2005 by Linus Torvalds, the famous creator of the Linux operating system kernel. A staggering number of software projects rely on Git for version control."

There are other VCS out there, but the most common way is to use git with github. 

### What is GitHub?
GitHub is a hosting service for Git repositories. It has most of the features of git in an online interface plus some others, like this "issue" that you're currently reading. We use github to:
Share your repositories with others.
Access other user's repositories.
Store remote copies of your repositories (on github's servers) as backup of your local copies.

### How does git work with GitHub?
This is the important part. 

Say you had a coding project with one file, "code.py". 

If you are working on this project alone, then you would have a folder on your computer that contains this file. The reason to use git here is to track changes to this file. So say you come to a point in your coding where you think you made a mistake and want to delete a big chunk of code. But you're not 100% sure this is the mistake, so after deleting that chunk of code you want to be able to get it back just in case. Without git, I would just make a copy of that file like "code_before_delete.py" and then just delete the code from "code.py" then see what happens. But this gets complicated in large projects or with code files where you have to make multiple decisions or deletions, you end up with a bunch of file copies, a bunch of commented out code, or code that you copy paste into a notepad file. It gets messy. git is the solution to that problem. It allows you to track every single change to every single file in a folder on your computer. SO you choose your project folder, tell git to start tracking changes here, and then you can add, delete files, change a line or a letter in a file, etc. and every time you want to create a "checkpoint" in your project, you tell git to save every change to all files, or only the changes to certain files as a specific "checkpoint". Git can show you the history of all your checkpoints, you can go back to a certain checkpoint and all files in your folder will magically be changed to how they were at that specific checkpoint. Pretty powerful stuff. Basically solves this (ignore "subversion" part):
![image](https://user-images.githubusercontent.com/40028187/57256177-f3ebe480-7023-11e9-90bd-f855f2e9087b.png)

Basically, for most software projects, you have many developers working on the same codebase (many different code files). SO you can have one developer write a code file, then save his changes. Now say another developer needs to use that code file, how does he know that the file has been updated

When I first started using github, I was like "ok this is great, but where do I actually code??"

**Comment by TyePie** [comment]
Very useful information and helps me understand what's going around on here.

Thank you!

**Comment by loucadufault** [comment]
hey happy to help!
